### ISSUE TITLE
Options for `postcss-rebase-url`

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [csstools/postcss-plugins#1315]
### What would you want to propose?

I'm currently using the following sequence of plugins:
```js
[
  postcssNesting(),
  postcssAssets({
    basePath: 'path/to/workspace',
    loadPaths: ['img'],
    relative: true,
  }),
  postcssPresetEnv(),
]
```
Here [postcss-assets](https://github.com/borodean/postcss-assets) turns `resolve('img/my-img.png')` into e.g. `url('../img/my-img.png')`.

Given that nesting is gaining native support and given that `postcss-assets` is no longer being developed, I'd like to switch over to [postcss-rebase-url](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-rebase-url) or maybe to [postcss-bundler](https://github.com/csstools/postcss-plugins/tree/main/plugin-packs/postcss-bundler) (which I guess would come *after* `postcss-preset-env` in the sequence).

However, `postcss-rebase-url` does not have the same flexibility as `postcss-assets`. If I'm reading the documentation correctly it currently expects URLs to be relative to the file path, whereas our URLs are relative to the `img` directory in the workspace root.

### Suggested solution

Extend `postcss-rebase-url` with some options:
* A base path option to let it resolve workspace-relative URLs and turn them into file-relative URLs
* A load paths option to resolve URLs relative to directories in the workspace (this isn't critical for my application as I can just add an explicit `img/` prefix)

If those options (or option) can be passed through via `@csstools/postcss-bundler` then I can probably switch to it.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/csstools/postcss-plugins/blob/master/CODE_OF_CONDUCT.md)
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### Would you like to open a PR for this feature?

- [ ] I'm willing to open a PR

**Comment by ehoogeveen-medweb** [comment]
Thanks for the reply :)

I can think of a few different potential uses for referencing files located at or near the root of the workspace:
* Shared images for e.g. `background` or `background-image`
* Shared stylesheets via `@import`
* Shared font files used via `src` in `@font-face`

Of those, I think shared images are the most compelling, but the idea is the same for all of them.

It is very common in JavaScript and TypeScript to refer to files relative to some specific root, with tooling support via e.g. tsconfig.json and browser support via import maps. I guess the tooling support isn't there for CSS, though a plugin to resolve and transform such URLs could warn if the referenced file can't be found.

**Comment by romainmenke** [comment]
> relative to some specific root

I think this is the largest unknown when rebasing urls.

`postcss-rebase-url` largely ignores this by taking the absolute path for the current file and resolving relative urls in CSS source against that.

So it has a very specific reference point (the current file) and this reference point has a knowable absolute path.

What would the reference point be for the suggested feature?
- current working directory
- absolute path of the top level stylesheet
- first ancestor directory with a `package.json` from the top level stylesheet
- ...

This reference point can then be used to resolve the plugin options against.
The resulting absolute paths can then again be used to rebase urls.

-------

Another strategy might be to have a callback function and leave it up to users to rewrite urls.

The callback function would be provided with arguments that give sufficient context, but the actual rewriting strategy would then be open.

**Comment by ehoogeveen-medweb** [comment]
Yes, a callback function could work; I currently have this for `postcss-import`:

```js
/**
 * Returns a custom resolver function using the given workspace path.
 *
 * @param {string} srcWorkspace - The absolute workspace path to resolve files relative to.
 */
function makeCustomResolver(srcWorkspace) {
	const srcWorkspaceCss = `${srcWorkspace}/css`;
	const srcWorkspaceJs = `${srcWorkspace}/js`;

	/**
	 * @param {string} id - The specified asset path to resolve.
	 * @param {string} basedir - The directory containing the active file.
	 * @returns {string} The resolved path or the empty string.
	 */
	function customResolver(id, basedir) {
		// Resolves CSS @imports relative to the parent directory, then css, then js, then the workspace.
		for (const path of [basedir, srcWorkspaceCss, srcWorkspaceJs, srcWorkspace]) {
			const resolved = nodePath.resolve(path, id).replaceAll('\\', '/');
			if (fileExists(resolved)) return resolved;
		}
		return '';
	}
	return customResolver;
}
```

Though admittedly it's a bit overkill considering how little we actually use it right now.

**Comment by romainmenke** [comment]
Hi @ehoogeveen-medweb,

I've prepared a draft change with a plugin that might be suitable : 
- https://github.com/csstools/postcss-plugins/pull/1317
- https://github.com/csstools/postcss-plugins/tree/1ddd5a004624e51a22fb4478a98cba9eb00042c2/plugins/postcss-rewrite-url#options

Please let me know if this covers your use case.

You still need to implement all of the actual path re-writing yourself but you will not need to think about gathering values, parsing, serializing, ...

**Comment by romainmenke** [comment]
🤔 Reading through this issue again I only now notice that you write a `resolve()` function with `postcss-assets`.

This is a good practice for non standard behavior, and I briefly forgot about that.

I think I will use `rewrite-url()`:
- same name as the plugin
- no conflict with standard CSS

**Comment by romainmenke** [comment]
> which I guess would come after `postcss-preset-env` in the sequence

Side note :

Your order should always be :
- plugins that bundle (`postcss-import`, `@csstools/postcss-bundler`, ...)
- plugins that remove non-standard CSS (sass like things, ...)
- plugins that polyfill or provide fallbacks (`postcss-preset-env`)
- plugins that minify (`cssnano`, `@csstools/postcss-minify`, ...)

It matters less with PostCSS 8 and later, but it still matters in some edge cases.

**Comment by ehoogeveen-medweb** [comment]
Thanks! Trying it out now and it seems to work fine :) I had a couple of remarks/questions:

1. What's the difference between `context.from` and `context.rootFrom`? When can they differ?
2. Would it be hard to make it work for `@import`s too? I'd presumably have to run the plugin twice (once before `postcss-import` or `@csstools/postcss-bundler`, and once after). Though to be honest I don't think it's a big deal if it doesn't work, it just seems more consistent to be able to transform `rewrite-url()` anywhere a `url()` might appear.

**Comment by ehoogeveen-medweb** [comment]
Oh and regarding the ordering, I guess I associated bundling with minifying and so assumed that it would come last, but I guess it's only for resolving/bundling imports.

**Comment by romainmenke** [comment]
> Though `@csstools/postcss-bundler` also doesn't know about `rewrite-url()`, so it wouldn't be able to rebase URLs in bundled files. Maybe this is opening up a can of worms that is better left closed.

Correct :)

A bundling plugin like `postcss-import` or `@csstools/postcss-bundler` would need to rebase `rewrite-url()` as an intermediate step. It needs to happen when parsing the styles to import. So that any `@import` statements found therein are also rebased. This is something `postcss-import` supports but `@csstools/postcss-bundler` does not.

I think this is a can of worms but I also like adding support for this, just because it is fairly trivial to do.

Then it is up to users of the plugin to use this in a way that actually works.


-----

> What's the difference between context.from and context.rootFrom? When can they differ?

> Aha, yes - in case of rewrite-url() in a file that gets @imported into another file, I want to use rootFrom so the URLs are made relative to the destination file (the bundle).

Yup, `rootFrom` is the "originating" or "root" file. `from` is the "current" file.

You might also use this pattern :

```
variables/theme-blue.css
variables/theme-green.css
components/card.css
components/button.css
```


```css
/* styles-blue.css */
@import "variables/theme-blue.css";
@import "components/card.css";
@import "components/button.css";
```

```css
/* styles-green.css */
@import "variables/theme-green.css";
@import "components/card.css";
@import "components/button.css";
```

**Comment by ehoogeveen-medweb** [comment]
Looks good!

One last thing I'm thinking about is whether it would be useful to have the option to keep `rewrite-url` (instead of changing it to `url`) for some sort of multi-pass transform. Of course this plugin is still the only one that's aware of what `rewrite-url` means, so I'm not sure whether there are any situations where a multi-pass transform would be needed (that can't be covered by existing properties like `from` and `rootFrom`).

Maybe another possibility would be to specify/override the function name so you could e.g. turn it into `resolve` and feed the result into `postcss-assets`, or something similar. I think should probably default to `url` though, unless you want to take "the plugin itself doesn't do any rewriting" to its logical extreme.

To be honest though, I don't have strong feelings about it as I don't think I would need it.

**Comment by romainmenke** [comment]
@ehoogeveen-medweb,

This was released : https://www.npmjs.com/package/@csstools/postcss-rewrite-url

------

> One last thing I'm thinking about is whether it would be useful to have the option to keep rewrite-url (instead of changing it to url) for some sort of multi-pass transform. 

I've added to option to return `false` in the callback.
This would then preserve the original `rewrite-url()`.

It doesn't give you the ability to transform the url while keeping the `rewrite-url` function name, but it does give you an escape hatch.
